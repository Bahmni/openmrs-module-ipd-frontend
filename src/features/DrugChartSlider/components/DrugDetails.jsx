import React from "react";
import { TextInput } from "carbon-components-react";
import {
  NumberInputCarbon,
  DropdownCarbon,
  DatePickerCarbon,
} from "bahmni-carbon-ui";
import PropTypes from "prop-types";
import moment from "moment";

export const DrugDetails = ({ hostData }) => {
  return (
    <>
      <TextInput
        id="drug-name"
        className="drug-name"
        type="text"
        value={hostData?.drugOrder?.drugOrder?.drug?.name}
        labelText="Drug Name"
        disabled
      />
      <div className="inline-field">
        <div className="dose-field-with-units">
          <NumberInputCarbon
            id={"Dropdown"}
            onChange={() => {}}
            style={{ width: "50%" }}
            label={"Dose"}
            value={hostData?.drugOrder?.uniformDosingType?.dose}
            isDisabled={true}
          />
          <DropdownCarbon
            id={"Dropdown"}
            onChange={() => {}}
            titleText={" "}
            style={{ paddingLeft: "10px" }}
            selectedValue={hostData?.drugOrder?.uniformDosingType?.doseUnits}
            options={[]}
            isDisabled={true}
          />
        </div>
        <div className="route">
          <DropdownCarbon
            id={"Dropdown"}
            onChange={() => {}}
            titleText={"Route"}
            selectedValue={hostData?.drugOrder?.route}
            options={[]}
            isDisabled={true}
          />
        </div>
      </div>
      <div className="inline-field">
        <div className="duration-field-with-units">
          <NumberInputCarbon
            id={"Dropdown"}
            onChange={() => {}}
            label={"Duration"}
            value={hostData?.drugOrder?.drugOrder?.duration}
            isDisabled={true}
          />
          <DropdownCarbon
            id={"Dropdown"}
            onChange={() => {}}
            titleText={" "}
            selectedValue={hostData?.drugOrder?.drugOrder?.durationUnits}
            options={[]}
            isDisabled={true}
          />
        </div>
        <DatePickerCarbon
          id={"Dropdown"}
          onChange={() => {}}
          titleText={"Start Date"}
          title={"Start Date"}
          value={moment(hostData?.drugOrder?.drugOrder?.scheduledDate).format(
            "MM/DD/YYYY"
          )}
          isDisabled={true}
        />
      </div>
      <div className="frequency">
        <DropdownCarbon
          id={"DropdownFrequency"}
          onChange={() => {}}
          titleText={"Frequency"}
          selectedValue={hostData?.drugOrder?.uniformDosingType?.frequency}
          options={[]}
          isDisabled={true}
        />
      </div>
    </>
  );
};

DrugDetails.propTypes = {
  hostData: PropTypes.shape({
    drugOrder: PropTypes.shape({
      drug: PropTypes.shape({
        name: PropTypes.string,
      }),
      uniformDosingType: PropTypes.shape({
        dose: PropTypes.string,
        doseUnits: PropTypes.string,
        frequency: PropTypes.string,
      }),
      route: PropTypes.string,
      duration: PropTypes.string,
      durationUnit: PropTypes.string,
      scheduledDate: PropTypes.string,
    }),
  }),
};
