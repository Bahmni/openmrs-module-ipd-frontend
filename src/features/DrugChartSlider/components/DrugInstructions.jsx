import React from "react";
import { TextArea } from "carbon-components-react";
import PropTypes from "prop-types";

export const DrugInstructions = ({ hostData }) => {
  return (
    <>
      <div className="instructions">
        <TextArea
          className="instruction"
          readOnly
          type="text"
          rows={1}
          value={hostData?.drugOrder?.instructions}
          labelText="Instruction"
          disabled
        />
      </div>
      <div className="additional-instructions">
        <TextArea
          className="additional-instruction"
          readOnly
          type="text"
          rows={1}
          value={hostData?.drugOrder?.additionalInstructions}
          labelText="Additional Instruction"
          disabled
        />
      </div>
    </>
  );
};

DrugInstructions.propTypes = {
  hostData: PropTypes.shape({
    drugOrder: PropTypes.shape({
      instructions: PropTypes.string,
      additionalInstructions: PropTypes.string,
    }),
  }),
};
