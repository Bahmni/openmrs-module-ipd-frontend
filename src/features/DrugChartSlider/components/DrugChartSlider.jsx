import React, { useContext, useEffect, useState } from "react";
import { TextArea } from "carbon-components-react";
import moment from "moment";
import PropTypes from "prop-types";
import { FormattedMessage } from "react-intl";
import { I18nProvider } from "../../i18n/I18nProvider";
import SideBarPanel from "../../SideBarPanel/components/SideBarPanel";
import "../styles/DrugChartSlider.scss";
import { medicationFrequency } from "../../../constants";
import { SaveAndCloseButtons } from "../../SaveAndCloseButtons/components/SaveAndCloseButtons";
import { SliderContext } from "../../../context/SliderContext";
import {
  isTimePassed,
  validateSchedules,
  saveMedication,
  updateStartTimeBasedOnFrequency,
  isInvalidTimeTextPresent,
  getUTCTimeEpoch,
} from "../utils/DrugChartSliderUtils";
import { DrugDetails } from "./DrugDetails";
import { DrugInstructions } from "./DrugInstructions";
import { StartTimeSection } from "./StartTimeSection";
import { ScheduleSection } from "./ScheduleSection";

const DrugChartSlider = (props) => {
  const { title, hostData, hostApi, setDrugChartNotes, drugChartNotes } = props;
  const enableSchedule = hostData?.scheduleFrequencies.find(
    (frequency) =>
      frequency.name === hostData?.drugOrder?.uniformDosingType?.frequency
  );
  const enableStartTime = hostData?.startTimeFrequencies.includes(
    hostData?.drugOrder?.uniformDosingType?.frequency
  );
  const enable24HourTimers = hostData?.enable24HourTimers || false;
  const isAutoFill = Boolean(enableSchedule?.scheduleTiming);

  const { setSliderContentModified } = useContext(SliderContext);
  const updateSliderContentModified = (value) => {
    setSliderContentModified((prev) => {
      return {
        ...prev,
        treatments: value,
      };
    });
  };
  const [firstDaySlotsMissed, setFirstDaySlotsMissed] = useState(0);
  const [startTime, setStartTime] = useState("");

  const [
    showStartTimeBeyondNextDoseWarning,
    setShowStartTimeBeyondNextDoseWarning,
  ] = useState(false);
  const [showStartTimePassedWarning, setShowStartTimePassedWarning] =
    useState(false);
  const [showEmptyStartTimeWarning, setShowEmptyStartTimeWarning] =
    useState(false);

  const [schedules, setSchedules] = useState([]);
  const [firstDaySchedules, setFirstDaySchedules] = useState([]);
  const [finalDaySchedules, setFinalDaySchedules] = useState([]);

  const [showSchedulePassedWarning, setShowSchedulePassedWarning] = useState(
    []
  );
  const [showScheduleOrderWarning, setShowScheduleOrderWarning] =
    useState(false);
  const [showEmptyScheduleWarning, setShowEmptyScheduleWarning] =
    useState(false);

  const [
    showFirstDaySchedulePassedWarning,
    setShowFirstDaySchedulePassedWarning,
  ] = useState([]);
  const [
    showFirstDayScheduleOrderWarning,
    setShowFirstDayScheduleOrderWarning,
  ] = useState(false);
  const [
    showEmptyFirstDayScheduleWarning,
    setShowEmptyFirstDayScheduleWarning,
  ] = useState(false);

  const [
    showFinalDayScheduleOrderWarning,
    setShowFinalDayScheduleOrderWarning,
  ] = useState(false);
  const [
    showEmptyFinalDayScheduleWarning,
    setShowEmptyFinalDayScheduleWarning,
  ] = useState(false);

  const handleFirstDaySchedule = (newSchedule, index) => {
    updateSliderContentModified(true);
    const newScheduleArray = [...firstDaySchedules];
    newScheduleArray[index] = enable24HourTimers
      ? newSchedule
      : moment(newSchedule, "hh:mm A");
    setFirstDaySchedules(newScheduleArray);
    if (!isInvalidTimeTextPresent(enable24HourTimers)) {
      setShowFirstDaySchedulePassedWarning((prevScheduleWarnings) => {
        const newSchedulePassedWarnings = [...prevScheduleWarnings];
        newSchedulePassedWarnings[index] = isTimePassed(newSchedule);
        return newSchedulePassedWarnings;
      });
    }
  };

  const handleSubsequentDaySchedule = (newSchedule, index) => {
    updateSliderContentModified(true);
    const newScheduleArray = [...schedules];
    newScheduleArray[index] = enable24HourTimers
      ? newSchedule
      : moment(newSchedule, "hh:mm A");
    setSchedules(newScheduleArray);
    if (!isInvalidTimeTextPresent(enable24HourTimers)) {
      setShowSchedulePassedWarning((prevScheduleWarnings) => {
        const newSchedulePassedWarnings = [...prevScheduleWarnings];
        newSchedulePassedWarnings[index] = isTimePassed(newSchedule);
        return newSchedulePassedWarnings;
      });
    }
  };

  const handleFinalDaySchedule = (newSchedule, index) => {
    updateSliderContentModified(true);
    const newScheduleArray = [...finalDaySchedules];
    newScheduleArray[index] = enable24HourTimers
      ? newSchedule
      : moment(newSchedule, "hh:mm A");
    setFinalDaySchedules(newScheduleArray);
  };

  const handleScheduleWarnings = async () => {
    const { isValid, warningType } = await validateSchedules(schedules);
    setShowEmptyScheduleWarning(!isValid && warningType === "empty");
    setShowScheduleOrderWarning(!isValid && warningType === "passed");
    return { isValid, warningType };
  };

  const isValidSchedule = async () => {
    const { isValid, warningType } = await handleScheduleWarnings();
    if (!isValid && (warningType === "empty" || warningType === "passed"))
      return false;
    return true;
  };

  const handleFirstDayScheduleWarnings = async () => {
    const { isValid, warningType } = await validateSchedules(
      firstDaySchedules.filter(
        (firstDaySchedule) => firstDaySchedule != "hh:mm"
      )
    );
    setShowEmptyFirstDayScheduleWarning(!isValid && warningType === "empty");
    setShowFirstDayScheduleOrderWarning(!isValid && warningType === "passed");
    return { isValid, warningType };
  };

  const isValidFirstDaySchedule = async () => {
    const { isValid, warningType } = await handleFirstDayScheduleWarnings();
    if (!isValid && (warningType === "empty" || warningType === "passed"))
      return false;
    return true;
  };

  const handleFinalDayScheduleWarnings = async () => {
    const { isValid, warningType } = await validateSchedules(finalDaySchedules);
    setShowEmptyFinalDayScheduleWarning(!isValid && warningType === "empty");
    setShowFinalDayScheduleOrderWarning(!isValid && warningType === "passed");
    return { isValid, warningType };
  };

  const isValidFinalDaySchedule = async () => {
    const { isValid, warningType } = await handleFinalDayScheduleWarnings();
    if (!isValid && (warningType === "empty" || warningType === "passed"))
      return false;
    return true;
  };

  const isStartTimeExceedingFrequency = (time, frequency) => {
    const currentTime = moment();
    const enteredTime = enable24HourTimers
      ? moment(time, "HH:mm")
      : moment(time, "hh:mm A");
    const updatedTime = updateStartTimeBasedOnFrequency(frequency, currentTime);
    return enteredTime.isAfter(updatedTime);
  };

  const handleStartTime = async (time) => {
    updateSliderContentModified(true);
    if (time === "") {
      setShowEmptyStartTimeWarning(true);
      setShowStartTimeBeyondNextDoseWarning(false);
      setShowStartTimePassedWarning(false);
      return;
    } else {
      setShowEmptyStartTimeWarning(false);
    }
    if (
      (enable24HourTimers && !moment(time, "HH:mm", true).isValid()) ||
      (!enable24HourTimers && !moment(time, "hh:mm A", true).isValid())
    ) {
      setStartTime(time);
      return;
    }
    isStartTimeExceedingFrequency(
      time,
      hostData?.drugOrder?.uniformDosingType?.frequency
    )
      ? setShowStartTimeBeyondNextDoseWarning(true)
      : setShowStartTimeBeyondNextDoseWarning(false);
    isTimePassed(time)
      ? setShowStartTimePassedWarning(true)
      : setShowStartTimePassedWarning(false);

    enable24HourTimers
      ? setStartTime(time)
      : setStartTime(moment(time, "hh:mm A"));
  };

  const createDrugChartPayload = () => {
    var payload = {
      providerUuid: hostData?.drugOrder?.provider?.uuid,
      patientUuid: hostData?.patientId,
      orderUuid: hostData?.drugOrder?.uuid,
      slotStartTime: null,
      firstDaySlotsStartTime: null,
      dayWiseSlotsStartTime: null,
      remainingDaySlotsStartTime: null,
      comments: drugChartNotes,
      medicationFrequency: "",
    };
    if (enableStartTime) {
      const startTimeUTCEpoch = getUTCTimeEpoch(
        startTime,
        enable24HourTimers,
        hostData?.drugOrder?.scheduledDate
      );
      payload.slotStartTime = startTimeUTCEpoch;
      payload.medicationFrequency =
        medicationFrequency.START_TIME_DURATION_FREQUENCY;
    }
    if (enableSchedule) {
      const nextScheduleDate = 24 * 60 * 60;
      const finalScheduleDate =
        nextScheduleDate * hostData?.drugOrder?.duration;

      const firstDaySchedulesUTCTimeEpoch = firstDaySchedules.reduce(
        (result, schedule) => {
          if (schedule !== "hh:mm") {
            result.push(
              getUTCTimeEpoch(
                schedule,
                enable24HourTimers,
                hostData?.drugOrder?.scheduledDate
              )
            );
          }
          return result;
        },
        []
      );

      const schedulesUTCTimeEpoch = schedules.map((schedule) =>
        getUTCTimeEpoch(
          schedule,
          enable24HourTimers,
          hostData?.drugOrder?.scheduledDate
        )
      );

      const finalDaySchedulesUTCTimeEpoch = finalDaySchedules.map((schedule) =>
        getUTCTimeEpoch(
          schedule,
          enable24HourTimers,
          hostData?.drugOrder?.scheduledDate
        )
      );

      payload.firstDaySlotsStartTime =
        firstDaySlotsMissed > 0 ? firstDaySchedulesUTCTimeEpoch : [];
      payload.dayWiseSlotsStartTime = firstDaySchedules.some(
        (schedule) => schedule == "hh:mm"
      )
        ? schedulesUTCTimeEpoch.map((schedules) => schedules + nextScheduleDate)
        : schedulesUTCTimeEpoch;
      const remainingDaySlotsStartTime = finalDaySchedulesUTCTimeEpoch.map(
        (schedules) => schedules + finalScheduleDate
      );

      const remainingDaySlotsTime = remainingDaySlotsStartTime.slice(
        0,
        firstDaySlotsMissed
      );
      payload.remainingDaySlotsStartTime = remainingDaySlotsTime;
      payload.medicationFrequency =
        medicationFrequency.FIXED_SCHEDULE_FREQUENCY;
    }
    return payload;
  };

  const validateSave = async () => {
    if (isInvalidTimeTextPresent(enable24HourTimers)) return false;
    if (enableSchedule) {
      const validFirstDaySchedules = await isValidFirstDaySchedule();
      const validSchedules = await isValidSchedule();
      const validFinalDaySchedules = await isValidFinalDaySchedule();
      if (!validFirstDaySchedules || !validSchedules || !validFinalDaySchedules)
        return false;
      return true;
    }
    if (enableStartTime) {
      if (!startTime) {
        setShowEmptyStartTimeWarning(true);
        return false;
      }
      return true;
    }
  };

  useEffect(() => {
    if (isAutoFill) {
      const scheduleTimings = enable24HourTimers
        ? enableSchedule?.scheduleTiming
        : enableSchedule?.scheduleTiming.map((time) => moment(time, "hh:mm A"));
      let finalScheduleCount = 0;
      scheduleTimings.forEach((schedule) => {
        if (isTimePassed(schedule)) {
          setFirstDaySchedules((prevSchedules) => [...prevSchedules, "hh:mm"]);
          finalScheduleCount = finalScheduleCount + 1;
          setFirstDaySlotsMissed((prevSlotNumber) => prevSlotNumber + 1);
        } else {
          setFirstDaySchedules((prevSchedules) => [...prevSchedules, schedule]);
        }
      });
      setSchedules(scheduleTimings || []);
      finalScheduleCount > 0 &&
        setFinalDaySchedules(scheduleTimings.slice(firstDaySlotsMissed) || []);
      if (finalScheduleCount == enableSchedule?.frequencyPerDay) {
        setFirstDaySchedules([]);
        setFinalDaySchedules([]);
        setFirstDaySlotsMissed(0);
      }
    } else {
      const defaultSchedules = Array.from(
        { length: enableSchedule?.frequencyPerDay },
        () => ""
      );
      setSchedules(defaultSchedules);
    }
  }, [isAutoFill, enable24HourTimers, enableSchedule]);

  const handleSave = async () => {
    const performSave = await validateSave();
    if (performSave) {
      const medication = createDrugChartPayload();
      const response = await saveMedication(medication);
      response.status === 200 ? hostApi.onModalSave?.() : null;
    }
  };

  const handleCancel = () => {
    hostApi.onModalCancel?.();
  };

  const handleClose = () => {
    hostApi.onModalClose?.();
  };

  const handleNotes = (e) => {
    updateSliderContentModified(true);
    setDrugChartNotes(e.target.value);
  };

  return (
    <I18nProvider>
      <SideBarPanel
        title={
          title ? title : <FormattedMessage id="DRUG_CHART_MODAL_HEADER" />
        }
        closeSideBar={handleClose}
      >
        <div style={{ padding: "20px", paddingBottom: "120px" }}>
          <DrugDetails hostData={hostData} />
          <ScheduleSection
            enableSchedule={enableSchedule}
            firstDaySlotsMissed={firstDaySlotsMissed}
            firstDaySchedules={firstDaySchedules}
            schedules={schedules}
            finalDaySchedules={finalDaySchedules}
            handleFirstDaySchedule={handleFirstDaySchedule}
            handleSubsequentDaySchedule={handleSubsequentDaySchedule}
            handleFinalDaySchedule={handleFinalDaySchedule}
            showFirstDayScheduleOrderWarning={showFirstDayScheduleOrderWarning}
            showEmptyFirstDayScheduleWarning={showEmptyFirstDayScheduleWarning}
            showFirstDaySchedulePassedWarning={
              showFirstDaySchedulePassedWarning
            }
            showScheduleOrderWarning={showScheduleOrderWarning}
            showEmptyScheduleWarning={showEmptyScheduleWarning}
            showFinalDayScheduleOrderWarning={showFinalDayScheduleOrderWarning}
            showEmptyFinalDayScheduleWarning={showEmptyFinalDayScheduleWarning}
            showSchedulePassedWarning={showSchedulePassedWarning}
            enable24HourTimers={enable24HourTimers}
          />
          {enableStartTime && (
            <StartTimeSection
              startTime={startTime}
              handleStartTime={handleStartTime}
              showEmptyStartTimeWarning={showEmptyStartTimeWarning}
              showStartTimeBeyondNextDoseWarning={
                showStartTimeBeyondNextDoseWarning
              }
              showStartTimePassedWarning={showStartTimePassedWarning}
              enable24HourTimers={enable24HourTimers}
            />
          )}
          <DrugInstructions hostData={hostData} />
          <div className="notes-sections">
            <TextArea
              data-testid="notes-section"
              className="notes-section"
              type="text"
              rows={3}
              value={drugChartNotes}
              onChange={(e) => handleNotes(e)}
              labelText="Notes"
            />
          </div>
        </div>
        <SaveAndCloseButtons
          onSave={() => handleSave()}
          onClose={() => handleCancel()}
        />
      </SideBarPanel>
    </I18nProvider>
  );
};

DrugChartSlider.propTypes = {
  title: PropTypes.string.isRequired,
  hostData: PropTypes.object.isRequired,
  hostApi: PropTypes.object.isRequired,
  setDrugChartNotes: PropTypes.func.isRequired,
  drugChartNotes: PropTypes.string.isRequired,
};

export default DrugChartSlider;
