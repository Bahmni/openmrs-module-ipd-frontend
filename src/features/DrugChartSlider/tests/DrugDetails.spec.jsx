import React from "react";
import { render } from "@testing-library/react";
import { DrugDetails } from "../components/DrugDetails";

const mockHostData = {
  drugOrder: {
    drug: {
      name: "drugName",
    },
    uniformDosingType: {
      dose: "dose",
      doseUnits: "doseUnits",
    },
    route: "route",
    duration: "duration",
  },
};

describe("DrugDetails", () => {
  it("should match snapshot", () => {
    const { container } = render(<DrugDetails hostData={mockHostData} />);
    expect(container).toMatchSnapshot();
  });
});
