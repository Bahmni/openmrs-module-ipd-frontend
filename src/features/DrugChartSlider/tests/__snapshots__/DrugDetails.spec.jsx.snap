// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrugDetails should match snapshot 1`] = `
<div>
  <div
    class="bx--form-item bx--text-input-wrapper"
  >
    <label
      class="bx--label bx--label--disabled"
      for="drug-name"
    >
      Drug Name
    </label>
    <div
      class="bx--text-input__field-outer-wrapper"
    >
      <div
        class="bx--text-input__field-wrapper"
      >
        <input
          class="bx--text-input drug-name bx--text-input--md"
          disabled=""
          id="drug-name"
          type="text"
          value="drugName"
        />
      </div>
    </div>
  </div>
  <div
    class="inline-field"
  >
    <div
      class="dose-field-with-units"
    >
      <div
        class="-TaKcdG2c0NUzkYaFWOzlQ=="
      >
        <div
          class="bx--form-item"
        >
          <div
            class="bx--number bx--number--helpertext"
          >
            <label
              class="bx--label bx--label--disabled"
              for="Dropdown"
            >
              <div
                class="titleText"
              >
                <span>
                  Dose 
                </span>
              </div>
            </label>
            <div
              class="bx--number__input-wrapper"
            >
              <input
                disabled=""
                id="Dropdown"
                pattern="[0-9]*"
                step="1"
                type="number"
                value="dose"
              />
              <div
                class="bx--number__controls"
              >
                <button
                  aria-label="Decrement number"
                  class="bx--number__control-btn down-icon"
                  disabled=""
                  tabindex="-1"
                  title="Decrement number"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="down-icon"
                    fill="currentColor"
                    focusable="false"
                    height="16"
                    preserveAspectRatio="xMidYMid meet"
                    viewBox="0 0 32 32"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M8 15H24V17H8z"
                    />
                  </svg>
                </button>
                <div
                  class="bx--number__rule-divider"
                />
                <button
                  aria-label="Increment number"
                  class="bx--number__control-btn up-icon"
                  disabled=""
                  tabindex="-1"
                  title="Increment number"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="up-icon"
                    fill="currentColor"
                    focusable="false"
                    height="16"
                    preserveAspectRatio="xMidYMid meet"
                    viewBox="0 0 32 32"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M17 15L17 8 15 8 15 15 8 15 8 17 15 17 15 24 17 24 17 17 24 17 24 15z"
                    />
                  </svg>
                </button>
                <div
                  class="bx--number__rule-divider"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="dropdown"
        data-testid="select dropdown"
        style="margin-right: 5px;"
      >
        <div
          class="bx--dropdown__wrapper bx--list-box__wrapper"
        >
          <label
            class="bx--label bx--label--disabled"
            for="downshift-0-toggle-button"
            id="downshift-0-label"
          >
            <div
              class="titleText"
            >
              <span>
                  
              </span>
            </div>
          </label>
          <div
            class="bx--dropdown bx--dropdown--disabled bx--list-box"
            id="Dropdown"
          >
            <button
              aria-disabled="true"
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="downshift-0-label downshift-0-toggle-button"
              class="bx--list-box__field"
              disabled=""
              id="downshift-0-toggle-button"
              title="doseUnits"
              type="button"
            >
              <span
                class="bx--list-box__label"
              >
                doseUnits
              </span>
              <div
                class="bx--list-box__menu-icon"
              >
                <svg
                  aria-label="Open menu"
                  fill="currentColor"
                  focusable="false"
                  height="16"
                  name="chevron--down"
                  preserveAspectRatio="xMidYMid meet"
                  role="img"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
                  />
                  <title>
                    Open menu
                  </title>
                </svg>
              </div>
            </button>
            <div
              aria-labelledby="downshift-0-label"
              class="bx--list-box__menu"
              id="downshift-0-menu"
              role="listbox"
              tabindex="-1"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="route"
    >
      <div
        class="dropdown"
        data-testid="select dropdown"
        style="margin-right: 5px;"
      >
        <div
          class="bx--dropdown__wrapper bx--list-box__wrapper"
        >
          <label
            class="bx--label bx--label--disabled"
            for="downshift-1-toggle-button"
            id="downshift-1-label"
          >
            <div
              class="titleText"
            >
              <span>
                Route 
              </span>
            </div>
          </label>
          <div
            class="bx--dropdown bx--dropdown--disabled bx--list-box"
            id="Dropdown"
          >
            <button
              aria-disabled="true"
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="downshift-1-label downshift-1-toggle-button"
              class="bx--list-box__field"
              disabled=""
              id="downshift-1-toggle-button"
              title="route"
              type="button"
            >
              <span
                class="bx--list-box__label"
              >
                route
              </span>
              <div
                class="bx--list-box__menu-icon"
              >
                <svg
                  aria-label="Open menu"
                  fill="currentColor"
                  focusable="false"
                  height="16"
                  name="chevron--down"
                  preserveAspectRatio="xMidYMid meet"
                  role="img"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
                  />
                  <title>
                    Open menu
                  </title>
                </svg>
              </div>
            </button>
            <div
              aria-labelledby="downshift-1-label"
              class="bx--list-box__menu"
              id="downshift-1-menu"
              role="listbox"
              tabindex="-1"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="inline-field"
  >
    <div
      class="duration-field-with-units"
    >
      <div
        class="-TaKcdG2c0NUzkYaFWOzlQ=="
      >
        <div
          class="bx--form-item"
        >
          <div
            class="bx--number bx--number--helpertext"
          >
            <label
              class="bx--label bx--label--disabled"
              for="Dropdown"
            >
              <div
                class="titleText"
              >
                <span>
                  Duration 
                </span>
              </div>
            </label>
            <div
              class="bx--number__input-wrapper"
            >
              <input
                disabled=""
                id="Dropdown"
                pattern="[0-9]*"
                step="1"
                type="number"
                value="duration"
              />
              <div
                class="bx--number__controls"
              >
                <button
                  aria-label="Decrement number"
                  class="bx--number__control-btn down-icon"
                  disabled=""
                  tabindex="-1"
                  title="Decrement number"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="down-icon"
                    fill="currentColor"
                    focusable="false"
                    height="16"
                    preserveAspectRatio="xMidYMid meet"
                    viewBox="0 0 32 32"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M8 15H24V17H8z"
                    />
                  </svg>
                </button>
                <div
                  class="bx--number__rule-divider"
                />
                <button
                  aria-label="Increment number"
                  class="bx--number__control-btn up-icon"
                  disabled=""
                  tabindex="-1"
                  title="Increment number"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="up-icon"
                    fill="currentColor"
                    focusable="false"
                    height="16"
                    preserveAspectRatio="xMidYMid meet"
                    viewBox="0 0 32 32"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M17 15L17 8 15 8 15 15 8 15 8 17 15 17 15 24 17 24 17 17 24 17 24 15z"
                    />
                  </svg>
                </button>
                <div
                  class="bx--number__rule-divider"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="dropdown"
        data-testid="select dropdown"
        style="margin-right: 5px;"
      >
        <div
          class="bx--dropdown__wrapper bx--list-box__wrapper"
        >
          <label
            class="bx--label bx--label--disabled"
            for="downshift-2-toggle-button"
            id="downshift-2-label"
          >
            <div
              class="titleText"
            >
              <span>
                  
              </span>
            </div>
          </label>
          <div
            class="bx--dropdown bx--dropdown--disabled bx--list-box"
            id="Dropdown"
          >
            <button
              aria-disabled="true"
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="downshift-2-label downshift-2-toggle-button"
              class="bx--list-box__field"
              disabled=""
              id="downshift-2-toggle-button"
              type="button"
            >
              <span
                class="bx--list-box__label"
              />
              <div
                class="bx--list-box__menu-icon"
              >
                <svg
                  aria-label="Open menu"
                  fill="currentColor"
                  focusable="false"
                  height="16"
                  name="chevron--down"
                  preserveAspectRatio="xMidYMid meet"
                  role="img"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
                  />
                  <title>
                    Open menu
                  </title>
                </svg>
              </div>
            </button>
            <div
              aria-labelledby="downshift-2-label"
              class="bx--list-box__menu"
              id="downshift-2-menu"
              role="listbox"
              tabindex="-1"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="date-picker-carbon"
      data-testid="datePicker"
    >
      <div
        class="bx--form-item"
      >
        <div
          class="bx--date-picker bx--date-picker--single"
          disabled=""
        >
          <div
            class="bx--date-picker-container"
          >
            <label
              class="bx--label bx--label--disabled"
              for="Dropdown"
            >
              <div
                class="titleText"
              >
                <span>
                  Start Date 
                </span>
              </div>
            </label>
            <div
              class="bx--date-picker-input__wrapper"
            >
              <input
                autocomplete="off"
                class="bx--date-picker__input bx--date-picker__input--md flatpickr-input"
                disabled=""
                id="Dropdown"
                pattern="\\d{1,2}\\/\\d{1,2}\\/\\d{4}"
                placeholder="mm/dd/yyyy"
                style="width: 250px;"
                type="text"
              />
              <svg
                aria-hidden="true"
                class="bx--date-picker__icon"
                fill="currentColor"
                focusable="false"
                height="16"
                preserveAspectRatio="xMidYMid meet"
                role="img"
                viewBox="0 0 32 32"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M26,4h-4V2h-2v2h-8V2h-2v2H6C4.9,4,4,4.9,4,6v20c0,1.1,0.9,2,2,2h20c1.1,0,2-0.9,2-2V6C28,4.9,27.1,4,26,4z M26,26H6V12h20	V26z M26,10H6V6h4v2h2V6h8v2h2V6h4V10z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="frequency"
  >
    <div
      class="dropdown"
      data-testid="select dropdown"
      style="margin-right: 5px;"
    >
      <div
        class="bx--dropdown__wrapper bx--list-box__wrapper"
      >
        <label
          class="bx--label bx--label--disabled"
          for="downshift-3-toggle-button"
          id="downshift-3-label"
        >
          <div
            class="titleText"
          >
            <span>
              Frequency 
            </span>
          </div>
        </label>
        <div
          class="bx--dropdown bx--dropdown--disabled bx--list-box"
          id="DropdownFrequency"
        >
          <button
            aria-disabled="true"
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="downshift-3-label downshift-3-toggle-button"
            class="bx--list-box__field"
            disabled=""
            id="downshift-3-toggle-button"
            type="button"
          >
            <span
              class="bx--list-box__label"
            />
            <div
              class="bx--list-box__menu-icon"
            >
              <svg
                aria-label="Open menu"
                fill="currentColor"
                focusable="false"
                height="16"
                name="chevron--down"
                preserveAspectRatio="xMidYMid meet"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
                />
                <title>
                  Open menu
                </title>
              </svg>
            </div>
          </button>
          <div
            aria-labelledby="downshift-3-label"
            class="bx--list-box__menu"
            id="downshift-3-menu"
            role="listbox"
            tabindex="-1"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
